Adds Block.getLogicalSize and limits size of response to client.

sshardool opened the pull request:
Resolves #11332
The change adds a method getLogicalSizeInBytes to the Block interface for
getting the size of a block's contents as if no optimizations (such as RLE or
Dictionary encoding) were applied to it.
This provides a more accurate estimate of the data size sent to the client when
Query::getNextResult is invoked.
For blocks other than RLE and Dictionary encoded the method returns getSizeInBytes().
More details at: https://github.com/prestodb/presto/issues/11332

sshardool commented:
Original changes for this were part of: #11350 
Review request: @electrum @dain @findepi

dain commented:
@haozhun, can you have a chat with @electrum about this one?  This has gone around twice already, and I'm afraid we are making @sshardool dizzy at this point.

sshardool commented:
@haozhun For reference these 2 PRs capture some of the discussions behind this:
#11572 
#11350

electrum commented on presto-main/src/test/java/com/facebook/presto/block/TestDictionaryBlock.java:
```diff
@@ -41,6 +41,22 @@ public void testSizeInBytes()
+        // The 10 Slices in the array will be of lengths 0 to 9.
+        Slice[] expectedValues = createExpectedValues(10);
+
+        // The dictionary within the dictionary block is expected to be a VariableWidthBlock of size 95bytes.
```
Nit: when I read "95bytes" it jumps out at me as missing a space

We could also write this as 95B, which would be consistent with values like 16MB, but that seems less natural than "95 bytes" (and could be confused with 95 billion).

sshardool replied:
Makes sense. I will fix this in all of my comments.

electrum commented on presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java:
```diff
@@ -163,18 +163,28 @@ default int compareTo(int leftPosition, int leftOffset, int leftLength, Block ri
 
     /**
-     * Returns the logical size of this block in memory.
+     * Returns the size of this block as if it was compacted, ignoring any over-allocations.
```
Let's add a second sentence to give a bit more clarity:
```
For example, in dictionary blocks, this only counts each dictionary entry once,
rather than each time a value is referenced.

sshardool replied:
Will update the comment based on  your suggestion.
Apart from that do you think the java doc should also reflect that for blocks returned by `Block.getRegion()` this will give the size of the 'view' instead of the size of the underlying block? I suspected it's already adequately captured by '_size of this block as if it was compacted_'.

electrum commented on presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java:
```diff
@@ -163,18 +163,28 @@ default int compareTo(int leftPosition, int leftOffset, int leftLength, Block ri
-     * Returns the logical size of {@code block.getRegion(position, length)} in memory.
+     * Returns the size of this block's contents as if no optimizations were applied to it.
+     * As example, for RLE and Dictionary blocks this can differ significantly from
+     * {@code block.getSizeInBytes()}.
```
I tried to expand on this. What do you think?
```
This can differ substantially from {@link #getSizeInBytes} for dictionary
or RLE blocks. For RLE, it will be {@code N} times larger. For dictionary,
it will be larger based on how many times dictionary entries reused.

sshardool replied:
Looks good. Will update the comment.

haozhun commented:
I was a bit concerned about potential misuse of the method (just see how it went for `getSizeInBytes`). I would say using it to estimate JSON serialized size is a misuse. However, I acknowledge that this is a step forward in making the target size better. As a result, I think this PR is reasonable. I do still believe the logic for serializing data for client output should eventually be migrated away from using `getLogicalSizeInBytes`.

I'm convinced that `getLogicalSizeInBytes` is a reasonable addition to Presto code, albeit not for this purpose.

sshardool commented:
@electrum I have incorporated your comments. 
-  Reverted the original commit and created 2 new ones based on the commit style doc.
-  Modified java doc and other comments.

@haozhun I agree, while this improves the current behavior this may not be the most ideal use of the new method. If a need arises for placing a more accurate limit on the serialized response in future then I would happy to look into the recommendations on #11572 which would apply the size limit after serialization.

electrum commented:
Merged, thanks!

sshardool commented on presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java:
```diff
@@ -219,6 +220,30 @@ private void calculateCompactSize()
+        if (logicalSizeInBytes >= 0) {
+            return logicalSizeInBytes;
+        }
+        logicalSizeInBytes = 0;
```
Makes sense. Thanks for taking care of it while merging.
